In [40]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [41]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df_lending_data = pd.read_csv("Resources/lending_data.csv")

# Review the DataFrame
df_lending_data.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [42]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending_data['loan_status']

# Separate the X variable, the features
X = df_lending_data.drop(columns=['loan_status'])

In [43]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [44]:
# Review the X variable DataFrame
X.sample(10)

loan_size  interest_rate  borrower_income  debt_to_income  \
35760    11900.0          8.202            57800        0.480969   
32755     9900.0          7.336            49600        0.395161   
52887    10000.0          7.364            49900        0.398798   
34863     9200.0          7.047            46900        0.360341   
15248    11700.0          8.083            56700        0.470899   
14317    10100.0          7.420            50400        0.404762   
72356     7500.0          6.292            39800        0.246231   
34569     9900.0          7.316            49400        0.392713   
31520    10400.0          7.553            51700        0.419729   
57320     7800.0          6.431            41100        0.270073   

       num_of_accounts  derogatory_marks  total_debt  
35760                6                 1       27800  
32755                4                 0       19600  
52887                4                 0       19900  
34863                3                 0       16900  
15248                5                 1       26700  
14317                4                 1       20400  
72356                2                 0        9800  
34569                4                 0       19400  
31520                4                 1       21700  
57320                2                 0       11100

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [45]:
# Check the balance of our target values
label_counts = y.value_counts()
print(label_counts)

0    75036
1     2500
Name: loan_status, dtype: int64


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [46]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [47]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(solver = 'lbfgs', max_iter = 200, random_state=1)

# Fit the model using training data
model.fit(X_train, y_train)

LogisticRegression(max_iter=200, random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [48]:
# Make a prediction using the testing data
y_pred = model.predict(X_test)
df_results = pd.DataFrame({
    "y_test": y_test,
    "pred": y_pred
})

df_results.sample(10)

y_test  pred
24863       0     0
74012       0     0
8119        0     0
42515       0     0
4239        0     0
41295       0     0
65989       0     0
56664       0     0
21181       0     0
50886       0     0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [49]:
# Print the balanced_accuracy score of the model
from sklearn.metrics import accuracy_score

# Calculate the accuracy score
accuracy = accuracy_score(y_test, y_pred)
accuracy= round(accuracy,3)
print("Accuracy:", accuracy)

Accuracy: 0.992


In [50]:
#Import the confusion_matrix module from SKLearn
from sklearn.metrics import confusion_matrix

# Generate a confusion matrix for the model
confusion_mtx = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(confusion_mtx)

Confusion Matrix:
[[18663   102]
 [   56   563]]


In [51]:
# Print the classification report for the model
from sklearn.metrics import classification_report

# Print the classification report
classification_rep = classification_report(y_test, y_pred)
print("Classification Report:")
print(classification_rep)

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:
In comparison to the original dataset, similarly the number of healthy loans is greater than the number of unhealthy loans. The model has a incredible model of 99%, the precision score for 0 (healthy loans) is 100% and the precision for 1 labels is okay at 85%. The recall score is also quite high at 99% for prediction of 0 labels and 91% for high-risk loans with the label 1.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [59]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ROS = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled= ROS.fit_resample(X_train, y_train)

In [60]:
# Count the distinct values of the resampled labels data
label_counts = y_resampled.value_counts()
num_distinct_labels = len(label_counts)

print("Number of distinct labels:", num_distinct_labels)
# print("Distinct labels and their counts:")
print(label_counts)

Number of distinct labels: 2
0    56271
1    56271
Name: loan_status, dtype: int64


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [61]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
from sklearn.linear_model import LogisticRegression
ros_model = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using the resampled training data
ros_model.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
y_pred = ros_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [62]:
# Print the balanced_accuracy score of the model 
print(f"Training Data Score: {ros_model.score(X_train, y_train)}")
print(f"Testing Data Score: {ros_model.score(X_test, y_test)}")

from sklearn.metrics import accuracy_score
# Display the accuracy score for the test dataset.
print(f"Balanced Accuracy Score: {balanced_accuracy_score(y_test, y_pred)}")

Training Data Score: 0.9945659650570917
Testing Data Score: 0.9938093272802311
Balanced Accuracy Score: 0.9936781215845847


In [63]:
# Generate a confusion matrix for the model
from sklearn.metrics import confusion_matrix
ros_testing_matrix = confusion_matrix(y_test, y_pred)
print(ros_testing_matrix)

[[18649   116]
 [    4   615]]


In [64]:
# Print the classification report for the model
ros_testing_report = classification_report(y_test, y_pred)

print(ros_testing_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer: The accuracy score for both prediction models is quite high at 99%. Looking at the confusion matrix, the oversampled data model did significantly better at predicting false negatives, meaning only 4 loans of 0 type were identified as false negative. Similar to the previous model, the precision score for 0 loans was 100% and 84% to loan type 1. The recall score improved for unhealthy loans compared to the previous model. Overall,the original model was quite accurate but this model with oversampled data seems to be slightly better at predicting more accurately.